executor/analyze.go

```diff
@@ -198,14 +155,88 @@ func (e *AnalyzeExec) Next(ctx context.Context, _ *chunk.Chunk) error {
 	if err != nil {
 		return err
 	}
+
+	// Update analyze options to mysql.analyze_options for auto analyze.
 	err = e.saveV2AnalyzeOpts()
 	if err != nil {
-		e.Ctx().GetSessionVars().StmtCtx.AppendWarning(err)
+		sessionVars.StmtCtx.AppendWarning(err)
+	}
+
+	if sessionVars.InRestrictedSQL {
+		return statsHandle.Update(infoSchema)
 	}
-	if e.Ctx().GetSessionVars().InRestrictedSQL {
-		return statsHandle.Update(e.Ctx().GetInfoSchema().(infoschema.InfoSchema))
+
+	return statsHandle.Update(infoSchema, cache.WithTableStatsByQuery())
+}
+
+// filterAndCollectTasks filters the tasks that are not locked and collects the table IDs.
+func filterAndCollectTasks(tasks []*analyzeTask, statsHandle *handle.Handle, infoSchema infoschema.InfoSchema) ([]*analyzeTask, uint, []string) {
+	var (
+		filteredTasks       []*analyzeTask
+		skippedTables       []string
+		needAnalyzeTableCnt uint
+		tids                = make(map[int64]struct{})
+	)
+
+	for _, task := range tasks {
+		tableID := getTableIDFromTask(task)
+		isLocked := statsHandle.IsTableLocked(tableID)
+		if !isLocked {
+			filteredTasks = append(filteredTasks, task)
+		}
+		if _, ok := tids[tableID]; !ok {
+			if isLocked {
+				tbl, ok := infoSchema.TableByID(tableID)
+				if !ok {
+					logutil.BgLogger().Warn("Unknown table ID in analyze task", zap.Int64("tid", tableID))
+					// Ignore this table because it may have been dropped.
+					continue
+				}
+				skippedTables = append(skippedTables, tbl.Meta().Name.L)
```

Comment:
```suggestion
				if !ok {
					logutil.BgLogger().Warn("Unknown table ID in analyze task", zap.Int64("tid", tableID))
				} else {
					skippedTables = append(skippedTables, tbl.Meta().Name.L)
				}
```
Looks like we shouldn't `continue` here because we still need to run `tids[tableID] = struct{}{}` below.

And the comment can be removed because dropping the table actually won't cause this.

Reply:
Addressed in [`4b587e8` (#46235)](https://github.com/pingcap/tidb/pull/46235/commits/4b587e8d07cdb82787c883aacaf0407fbb2ac54b)